"comparator" option for properties

schnepe2 opened the pull request:
My proposal for the feature request #9039.
Allowed variations for "`comparator`" are
-  `comparator : "a==b" // String` (inline) parameter signature bool::(var a, var b)
-  `comparator : function (a,b) { return a===b; } // Function`
-  `comparator : "__checkIt" // String` (member function)

schnepe2 committed:
Update Property.js

johnspackman commented on framework/source/class/qx/core/Property.js:
```diff
@@ -891,6 +894,48 @@ qx.Bootstrap.define("qx.core.Property",
+      {
+        var members = clazz.prototype;
+        // Name of member?
+        if (members[config.comparator]!==undefined)
```
what if the member is in a base class?

level420 replied:
So in addition to the first idea the string type comparator does not only contain some simple code which is wrapped in a function, but may also contain a reference to a class method by its name.
I think we should ommit this part and only have the inline code,
If we we call the inline code wrapped function in the context of the instance, we have access to all instance methods via ```this.```

schnepe2 replied:
> does this work if the method is defined in a superclass?

It seem so (last context-changes might be necessary for it)
test case is in work...

level420 commented:
@schnepe2 Thank you for the PR! It would be great if you also can create a unit test for the new functionality. At best one for the string and one for the function type of comparator. That would be great!

level420 commented on framework/source/class/qx/core/Property.js:
```diff
@@ -991,7 +1036,7 @@ qx.Bootstrap.define("qx.core.Property",
       if (incomingValue) {
-        code.push('if(this.', store, '===value)return value;');
+        code.push('if(equ(this.', store, ',value))return value;');
       } else if (resetValue) {
```
Shouldn't this be called in the instance context?

```javascript
code.push('if(equ.call(this, this.', store, ',value))return value;');
```

schnepe2 replied:
...not sure, should it? It worked for me.
But please write a test-case which shows the potential problem

schnepe2 replied:
Ah! I see! and **yes** it **should** be called with the instances context!
Thanks for the hint

schnepe2 commented:
> @schnepe2 Thank you for the PR! It would be great if you also consider to create a unit test for the new functionality. At best one for the string and one for the function type of comparator. That would be very helpfull and would make your PR complete!

I am already on it (I would never commit any change without a test :smile: !), but as I said, I am inexperienced with git / git-hub / ...

schnepe2 committed:
Added Unit Test(s)

Added unit tests for #9039 feature

level420 commented on framework/source/class/qx/test/core/Property.js:
```diff
@@ -570,5 +570,170 @@ qx.Class.define("qx.test.core.Property",
 
       object.dispose();
     }
+
```
@schnepe2 you missed a semicolon here

schnepe2 committed:
fixed syntax error (missing semicolon)

..the disadvantages of using the web interface! You cut 100% but paste only 99.9999...

schnepe2 committed:
Added context checking tests

schnepe2 committed:
fixed context in "comparator" calls

schnepe2 committed:
Added base class member call test

schnepe2 commented:
I have a general question regarding throwing the exception (around line 930):
 ```
else
{
  throw new Error( "Invalid type for 'comparator' attribute " +
                   "of property '" + name + "' in class '" + clazz.classname + "'" );
}
```
should this be encapsulated in "qx.debug" definition?
 ```
else if (qx.core.Environment.get("qx.debug"))
{
  throw new Error( "Invalid type for 'comparator' attribute " +
                   "of property '" + name + "' in class '" + clazz.classname + "'" );
}
```
I think it's enough to check only in source builds ("qx.debug").
Depending on your decision I will add test-cases for this code, too

schnepe2 committed:
Extended property "comparator" Tests for invalid values

derrell commented on framework/source/class/qx/core/Property.js:
```diff
@@ -272,7 +272,8 @@ qx.Bootstrap.define("qx.core.Property",
       deferredInit : "boolean",  // Boolean
-      validate     : null        // String, Function
+      validate     : null,       // String, Function
+      comparator   : null        // String, Function
```
Can we change this to a name that implies the meaning of the return value? Instead of comparator, how about isEqual or something of the sort?

derrell commented:
Please change "comparator" to "isEqual" so that the property name implies the return value.

level420 commented:
I think that comparator is a good choice. Of course something like ```compareLogikForTestingTheEqualityOfOldAndNewValue``` would be best, but that is too much ;-)
In the end you need to read the documentation to really know what the attribute does.
The attribute ```event``` does also not completely self explain its purpose by its name.

cajus commented:
I can't really vote it up or down as I don't understand - or never had - the usecase. As there's much discussion on that it seems to be there.

Just from the reading, I don't understand what `comparator` means and what it is intended for. I guess a word should raise an idea of what it does in the developers mind. From the wording it feels like something that helps comparing something with something else. But what?

In fact it is applied as a check if a value has changed. Maybe just the default way to compare values is not strict enough?

Tends to break stuff if we change that. What about naming it `valueChangedCheck`? Sure, it's a bit long, but it goes conform with the `check` naming and it expresses what it is used for. And: as I didn't had the need for it for the last 10 years, it might not be used so often to get angry about the property length ;-)

level420 commented:
The use case is where a simple comparison ```a === b``` fails. But yes we can name it ```valueChangedCheck```.

level420 commented:
vote for ```comparator``` by adding a thumbs up emoji to this comment

level420 commented:
vote for ```isEqual``` by adding a thumbs up emoji to this comment

level420 commented:
vote for ```valueChangedCheck``` by adding a thumbs up emoji to this comment

derrell commented:
I think maybe I didn't make the point I was trying to make, clearly. I am pushing for boolean function names that imply the meaning of the result. When you see the name *comparator* and it returns true, you don't know for sure what what that means. Similarly, if you see the name *valueChangedCheck*, you aren't sure.

On the other hand, if you see *isEqual* and it returns true, you know, without consulting the documentation, that the arguments are considered equal. The same could be accomplished with *valueChanged* (as opposed to *valueChangedCheck*). With *valueChanged*, you know that if it returns true, the value has changed.

My goal here is to lean towards an API in which one needn't always be consulting the docs to know the meaning of return values (or arguments, in other cases).

johnspackman commented:
I think we're spending way too much time debating this - any of those names, IMHO, are just as good as the other, and whether you're looking at `apply`, `check`, `transform` or `comparator`/`isEqual`/`valueChanged*` you would have to understand the API to know how to use it

What is important is that the name is concise, easily remembered, and doesn't have any obvious overlap with other concepts.

Lets just pick one and move on

oetiker commented:
Since we are developing a framework, I think it is worth to discuss these things ... it is the attention to detail that makes it a joy to work with qx ... 

And since no one seems to have any point AGAINST calling it something that makes sense, lets follow @derrell  s advise ...

zaucker commented:
On Mon, 13 Jun 2016, Tobias Oetiker wrote:

> Since we are developing a framework, I think it is worth to discuss these
> things ...  it is the attention to detail that makes it a joy to work with
> qx ...

+1

> And since no one seems to have any point AGAINST calling it something that
> makes sense, lets follow @derrell  s advise ...

+1

Cheers,
Fritz

-- 
Oetiker+Partner AG              tel: +41 62 775 9903 (direct)
Fritz Zaucker                        +41 62 775 9900 (switch board)
Aarweg 15                            +41 79 675 0630 (mobile)
CH-4600 Olten                   fax: +41 62 775 9905
Schweiz                         web: www.oetiker.ch

level420 commented:
Hmm. So do we have a consent?

slachtar commented:
upvote for isEqual()

On Mon, Jun 13, 2016 at 1:27 PM, Dietrich Streifert <
notifications@github.com> wrote:

> Hmm. So do we have a consent?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qooxdoo/qooxdoo/pull/9040#issuecomment-225566797>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/ACaj6n8JPqbvs8trC2bnlwjyVnmk95-rks5qLUzEgaJpZM4Ix9Ud>
> .
>



-- 
Slah Lachtar
Tel. (+216) 98 221 575

level420 commented:
We have the following vote:

- ```isEqual``` : 3
- ```valueChangedCheck``` : 5
- ```comparator``` : 1

where @oetiker voted double for isEqual and valueChangedCheck.

cajus commented:
Maybe it should be reset because @derrell 's arguments came in later? ;-)

level420 commented:
OK then. here we go again:

level420 commented:
vote for ```comparator``` by adding a thumbs up emoji to this comment

level420 commented:
vote for ```isEqual``` by adding a thumbs up emoji to this comment

level420 commented:
vote for ```valueChangedCheck``` by adding a thumbs up emoji to this comment

level420 commented:
@schnepe2 By voting and by final decission of @derrell : Please rework the PR to use ```isEqual``` as the attributes name instead of ```comparator```. Thank you for your help and effort.

level420 commented:
Voting ist closed ;-)

schnepe2 committed:
Rename of 'comparator' to 'isEqual'

schnepe2 committed:
Rename of 'comparator' to 'isEqual'

schnepe2 commented:
> [...] Please rework the PR to use isEqual as the attributes name instead of comparator.

Done!

derrell commented on framework/source/class/qx/core/Property.js:
```diff
@@ -891,6 +894,48 @@ qx.Bootstrap.define("qx.core.Property",
+      {
+        var members = clazz.prototype;
+        // Name of member?
+        if (members[config.isEqual]!==undefined)
```
Being picky here. For consistency with line 926, should this be changed to use typeof?
```
if (typeof members[config.isEqual] !== "undefined")
```

level420 replied:
What is the difference?

derrell replied:
members[config.isEqual] !== "undefined"
vs
typeof members[config.isEqual] !== "undefined"

The two do the same thing, but both are used in the same code, just a few lines apart, that's all.

derrell replied:
Y'know what... too picky. Forget about it. It's fine.

level420 commented:
Merging now. 

@schnepe2 Thank you for your work!
